flow/layers/image_filter_layer.cc

```diff
@@ -67,42 +73,10 @@ void ImageFilterLayer::Preroll(PrerollContext* context,
 
   set_paint_bounds(child_bounds);
 
-  SkMatrix child_matrix(matrix);
-#ifndef SUPPORT_FRACTIONAL_TRANSLATION
-  child_matrix = RasterCache::GetIntegralTransCTM(child_matrix);
-#endif
-
   transformed_filter_ = nullptr;
-  if (render_count_ >= kMinimumRendersBeforeCachingFilterLayer) {
-    // We have rendered this same ImageFilterLayer object enough
-    // times to consider its properties and children to be stable
-    // from frame to frame so we try to cache the layer itself
-    // for maximum performance.
-    TryToPrepareRasterCache(context, this, child_matrix,
-                            RasterCacheLayerStrategy::kLayer);
-  } else {
-    // This ImageFilterLayer is not yet considered stable so we
-    // increment the count to measure how many times it has been
-    // seen from frame to frame.
-    render_count_++;
-
-    // Now we will try to pre-render the children into the cache.
-    // To apply the filter to pre-rendered children, we must first
-    // modify the filter to be aware of the transform under which
-    // the cached bitmap was produced. Some SkImageFilter
-    // instances can do this operation on some transforms and some
-    // (filters or transforms) cannot. We can only cache the children
-    // and apply the filter on the fly if this operation succeeds.
-    transformed_filter_ = filter_->makeWithLocalMatrix(child_matrix);
-    if (transformed_filter_) {
-      // With a modified SkImageFilter we can now try to cache the
-      // children to avoid their rendering costs if they remain
-      // stable between frames and also avoiding a rendering surface
-      // switch during the Paint phase even if they are not stable.
-      // This benefit is seen most during animations.
-      TryToPrepareRasterCache(context, this, matrix,
-                              RasterCacheLayerStrategy::kLayerChildren);
-    }
+  transformed_filter_ = filter_->makeWithLocalMatrix(matrix);
+  if (transformed_filter_) {
+    layer_raster_cache_item_->MarkCacheChildren();
```

Comment:
Everytime when we entry the `LayerRasterItem::PrerollSetup` we will set the `cache_state_ = kNone`
```C++

void LayerRasterCacheItem::PrerollSetup(PrerollContext* context,
                                        const SkMatrix& matrix) {
  cache_state_ = CacheState::kNone;
 ...
}
```
I think if the `transformed_filter_` is `null`, the `cache_state` is still `kNone`, we don't cache the layer's children. Even thought we set the flag is `kChildren` in last frame.

Reply:
I disagree. transformed_filter_ being null has nothing to do with the cache item being in any state. It will be null if Skia can't handle the operation and for no other reason.

if the transform is the reason that Skia can't handle the operation then the status of whether transformed_filter_ is null could change any time the matrix changes and yet this code will only ever set "can_cache_children_" to true and never return it to false.

Reply:
If we can create a transformed filter then we can cache the children, otherwise we cannot. The ability to create the transformed filter can vary frame to frame independently of other factors about whether or not we can cache the layer.

Reply:
Yes, I know the `transformed_filter_` if is null and the layer render_count is less than 3, we will cache nothing, if the `transformed_filter_` isn't null and the layer render_count is less than 3, we will cache it children so if here will set the cache state is `CahceChildren`;
In layer_raster_cache_item first we will check the render_count if greate than 3, if so we well cache the layer
```C++
void LayerRasterCacheItem::PrerollFinalize(PrerollContext* context,
                                           const SkMatrix& matrix) {
  ....
  if (num_cache_attempts_ >= layer_cached_threshold_) {
    // the layer can be cached
    cache_state_ = CacheState::kCurrent;
    context->raster_cache->MarkSeen(key_id_, matrix_);
  } else {
    num_cache_attempts_++;
   ....
}
```
if the transformed_filter is not null and render_count less than 3, we cache children
```C++
void LayerRasterCacheItem::PrerollFinalize(PrerollContext* context,
                                           const SkMatrix& matrix) {
  if (num_cache_attempts_ >= layer_cached_threshold_) {
    // the layer can be cached
 ...
  } else {
    num_cache_attempts_++;
    // access current layer
    if (can_cache_children_) {
...
    }
  }
}
```
if the layer is not match any branch, we will cache none.

So for ImageFilterLayer if any logic which is I miss?

Reply:
The creation of the transformed_filter can succeed or fail independently each time Preroll is called. The success on one frame has nothing to do with the success on a subsequent frame. It needs to be checked *every* frame and the value of "can_cache_children" must match the success for the current frame. On some frames it will be true, on others it will be false and can go back and forth between true and false as many times as it exists in the layer tree.

LayerCacheItem can only set the value to false in the initializer and then it can only be set to true after that by using "MarkCacheChildren". There is no way to set the value to false once it is set to true, but if a subsequent frame fails to create the transformed filter, then we need a way to reflect that by setting can_cache_children to false. How do we do that?

Reply:
Aha, Sorry, It's my mistake